File: lib/src/main/java/com/king/frame/mvvmframe/di/module/HttpModule.java
package com.king.frame.mvvmframe.di.module;


import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.king.frame.mvvmframe.config.AppliesOptions;
import com.king.frame.mvvmframe.http.InterceptorConfig;
import com.king.frame.mvvmframe.http.interceptor.LogInterceptor;
import com.king.retrofit.retrofithelper.RetrofitHelper;

import javax.inject.Singleton;

import androidx.annotation.Nullable;
import dagger.Module;
import dagger.Provides;
import dagger.hilt.InstallIn;
import dagger.hilt.components.SingletonComponent;
import okhttp3.HttpUrl;
import okhttp3.OkHttpClient;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * @author <a href="mailto:jenly1314@gmail.com">Jenly</a>
 */
@InstallIn(SingletonComponent.class)
@Module
public class HttpModule {


    @Singleton
    @Provides
    Retrofit provideRetrofit(Retrofit.Builder builder,@Nullable AppliesOptions.RetrofitOptions options){
        if(options != null){
            options.applyOptions(builder);
        }
        return builder.build();
    }

    @Singleton
    @Provides
    OkHttpClient provideOkHttpClient(OkHttpClient.Builder builder,@Nullable AppliesOptions.OkHttpClientOptions options){
        if(options != null) {
            options.applyOptions(builder);
        }
        return builder.build();
    }

    @Singleton
    @Provides
    Gson provideGson(GsonBuilder builder,@Nullable AppliesOptions.GsonOptions options){
        if(options != null){
            options.applyOptions(builder);
        }
        return builder.create();
    }

    @Singleton
    @Provides
    Retrofit.Builder provideRetrofitBuilder(HttpUrl httpUrl, OkHttpClient client, Gson gson,InterceptorConfig config) {
        Retrofit.Builder builder = new Retrofit.Builder();
            builder.baseUrl(httpUrl)
                    .client(client);
            if(config.isAddGsonConverterFactory()){
                builder.addConverterFactory(GsonConverterFactory.create(gson));
            }

        return builder;
    }

    @Singleton
    @Provides
    OkHttpClient.Builder provideClientBuilder(InterceptorConfig config) {
        OkHttpClient.Builder builder = RetrofitHelper.getInstance().createClientBuilder();

        if(config.isAddLog()){
            builder.addInterceptor(new LogInterceptor());
        }

        return builder;
    }

    @Singleton
    @Provides
    GsonBuilder provideGsonBuilder(){
        return new GsonBuilder();
    }

    @Singleton
    @Provides
    InterceptorConfig provideInterceptorConfig(InterceptorConfig.Builder builder, @Nullable AppliesOptions.InterceptorConfigOptions options){
        if(options != null){
            options.applyOptions(builder);
        }
        return builder.build();
    }

    @Singleton
    @Provides
    InterceptorConfig.Builder provideInterceptorConfigBuilder(){
        return InterceptorConfig.newBuilder();
    }


}

File: mvvmframe/src/main/java/com/king/frame/mvvmframe/di/module/HttpModule.kt
package com.king.frame.mvvmframe.di.module

import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.king.frame.mvvmframe.config.AppliesOptions
import com.king.frame.mvvmframe.config.AppliesOptions.GsonOptions
import com.king.frame.mvvmframe.config.AppliesOptions.OkHttpClientOptions
import com.king.frame.mvvmframe.config.AppliesOptions.RetrofitOptions
import com.king.frame.mvvmframe.config.Config
import com.king.retrofit.retrofithelper.RetrofitHelper
import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.components.SingletonComponent
import okhttp3.HttpUrl
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import timber.log.Timber
import javax.inject.Singleton

/**
 * Http相关注入
 *
 * @author <a href="mailto:jenly1314@gmail.com">Jenly</a>
 * <p>
 * <a href="https://github.com/jenly1314">Follow me</a>
 */
@InstallIn(SingletonComponent::class)
@Module
object HttpModule {

    @Singleton
    @Provides
    fun provideRetrofit(builder: Retrofit.Builder, options: RetrofitOptions?): Retrofit {
        options?.applyOptions(builder)
        return builder.build()
    }

    @Singleton
    @Provides
    fun provideOkHttpClient(
        builder: OkHttpClient.Builder,
        config: Config,
        options: OkHttpClientOptions?
    ): OkHttpClient {
        options?.applyOptions(builder)
        val loggingInterceptor = HttpLoggingInterceptor().apply {
            level = config.httpLoggingLevel
        }
        builder.addInterceptor(loggingInterceptor)
        return builder.build()
    }

    @Singleton
    @Provides
    fun provideGson(builder: GsonBuilder, options: GsonOptions?): Gson {
        options?.applyOptions(builder)
        return builder.create()
    }

    @Singleton
    @Provides
    fun provideConfig(
        builder: Config.Builder,
        options: AppliesOptions.ConfigOptions?
    ): Config {
        options?.applyOptions(builder)
        return builder.build()
    }

    @Singleton
    @Provides
    fun provideRetrofitBuilder(
        httpUrl: HttpUrl,
        client: OkHttpClient,
        gson: Gson,
        config: Config
    ): Retrofit.Builder {
        val builder = Retrofit.Builder()
        builder.baseUrl(httpUrl)
            .client(client)
        if (config.isAddGsonConverterFactory) {
            builder.addConverterFactory(GsonConverterFactory.create(gson))
        }
        return builder
    }

    @Singleton
    @Provides
    fun provideClientBuilder(): OkHttpClient.Builder {
        return RetrofitHelper.getInstance().createClientBuilder()
    }

    @Singleton
    @Provides
    fun provideGsonBuilder(): GsonBuilder {
        return GsonBuilder()
    }

    @Singleton
    @Provides
    fun provideConfigBuilder(): Config.Builder {
        return Config.Builder()
    }

    private const val TAG = "OkHttp"
}
